torch/_jit_internal.py

```diff
@@ -97,16 +102,25 @@ def env(key):
 def weak_script(fn, _frames_up=0):
     """
     Marks a function as a weak script function. When used in a script function
-    or ScriptModule, the weak script function will be lazily compiled and
-    inlined in the graph. When not used in a script function, the weak script
-    annotation has no effect.
+    or ScriptModule or when tracing, the weak script function will be lazily
+    compiled and inlined in the graph. Otherwise, the weak script annotation
+    has no effect.
     """
     compiled_weak_fns[fn] = {
         "status": COMPILATION_PENDING,
         "compiled_fn": None,
         "rcb": createResolutionCallback(_frames_up + 1)
     }
-    return fn
+
+    def wrap_fn(*args, **kwargs):
+        if _get_tracing_state():
+            compiled_fn = _try_compile_weak_script(fn)
+            return compiled_fn(*args, **kwargs)
+        else:
+            return fn(*args, **kwargs)
+
+    weak_wrap_to_fn[wrap_fn] = fn
```

Comment:
This issue is that when you're compiling from ScriptModule, you only have access to the wrap function, and within wrap function you only have access to the inner fn. So mapping from wrap -> inner fixes the problem. I probably dont need any changes to boolean_dispatch with functools since the doc / name will propagate